lib/SILOptimizer/Utils/LexicalDestroyFolding.cpp

```diff
@@ -0,0 +1,772 @@
+  ///       end_borrow %lifetime
+  ///       ...
+  ///       destroy_value %original_borrowee
+  ///   to
+  ///       end_borrow %lifetime
+  ///       apply %fn(..., %move, ...)
+  ///       ...
+  /// - update SSA now that a second def has been introduced for
+  ///   %original_borrowee
+  void rewrite(SmallVectorImpl<Candidate> const &candidates);
+
+  /// Add the new move_value [lexical] above the begin_borrow [lexical].
+  ///
+  /// At most one will be created per run of LexicalDestroyFolding.
+  ///
+  /// returns false if the move_value [lexical] instruction was added already
+  ///         true if the move_value [lexical] instruction was added just now
+  bool createMove();
+
+  /// Combine the matched instruction sequence
+  ///
+  ///   apply %fn(%copy)
+  ///   end_borrow %lifetime
+  ///   destroy_value %instance
+  ///
+  /// into
+  ///
+  ///   end_borrow %lifetime
+  ///   apply %fn(%move)
+  ///
+  /// This is done in three steps:
+  /// (1) rewrite the apply (and delete the copy_value that is fed to it if
+  ///     possible)
+  /// (2) hoist the end_borrow
+  /// (3) delete the destroy_value
+  void fold(Match, SmallVectorImpl<int> const &rewritableArgumentIndices);
+
+  // Update SSA to reflect that fact that %move and %original_borrowee are two
+  // defs for the "same value".
+  void updateSSA();
+
+  /// The number of total uses of the simple extended borrow scope defined by
+  /// %introducer used as arguments in the specified apply instruction.
+  ///
+  /// Just does a lookup into extendedSimpleUseCountPerApplySite which is
+  /// calculated during findUsers.
+  size_t extendedSimpleUseCount(ApplyInst *ai) const;
+
+  /// Whether the specified value is could be replaced with %original_borrowee.
+  ///
+  /// In order to be, it has to be in the simple extended def of
+  /// %original_borrowee.
+  bool canReplaceValueWithBorrowee(SILValue value) const;
+
+  /// Find the arguments in the specified apply that could be rewritten.
+  void rewritableArgumentIndicesForApply(ApplySite,
+                                         SmallVectorImpl<int> &indices) const;
+
+private:
+  // State calculated over the course of the run.
+
+  // The operands that are uses of the original borrowee.
+  //
+  // Populated in findDominatedUsesOfOriginalBorrowee.
+  SmallVector<Operand *, 16> originalBorroweeUses;
+  // The instructions that are users of the original borrowee.
+  //
+  // A set of the users of originalBorroweeUses for fast membership checking.
+  //
+  // Populated in findDominatedUsesOfOriginalBorrowee.
+  SmallPtrSet<SILInstruction *, 16> originalBorroweeUsers;
+  // The instructions that are users of the simple extended borrow scope.
+  //
+  // Populated during findUsers.
+  SmallPtrSet<SILInstruction *, 16> users;
+  // Per apply site, the number of owned, simple extended uses of the borrow.
+  //
+  // Defined during findUsers.
+  llvm::SmallDenseMap<ApplySite, size_t> extendedSimpleUseCountPerApplySite;
```

Comment:
`extendedSimpleUseCountPerApplySite` is complicated. What happens if you just keep track of the one consuming argument that replaces the destroy and let updateSSA rewrite the operands of the other copied arguments (guaranteed arguments that use the borrowed value still aren't allowed)

Reply:
`extendedSimpleUseCountPerApplySite` has been eliminated by checking whether each operands of the apply in rewritableArgumentIndicesForApply is a use.  If it's a use but the wrong kind (i.e. not owned or not a copy of `%introducer`), the function returns false.  I don't think updateSSA as written would handle rewriting the `copy_value %introducer` instructions as `copy_value %move` for other copies of `%introducer` than the first.